rest-api/offline/public_metrics_export.py

```diff
@@ -0,0 +1,145 @@
+  # "PMI_".
+  ('gender',
+   """
+   SELECT code.value, ps.count
+   FROM (
+    SELECT gender_identity_id, SUM(1) count
+    FROM participant_summary
+    WHERE gender_identity_id IS NOT NULL
+    GROUP BY 1
+   ) ps LEFT JOIN code
+   ON ps.gender_identity_id = code.code_id;
+   """,
+   None),
+  ('race',
+   """
+   SELECT race, SUM(1)
+   FROM participant_summary
+   WHERE race IS NOT NULL
+   GROUP BY 1
+   """,
+   lambda v: Race.lookup_by_number(v).name),
+  ('state',
+   """
+   SELECT SUBSTR(code.value, {}), ps.count
+   FROM (
+    SELECT state_id, SUM(1) count
+    FROM participant_summary
+    WHERE state_id IS NOT NULL
+    GROUP BY 1
+   ) ps LEFT JOIN code
+   ON ps.state_id = code.code_id;
+   """.format(len('PIIState_') + 1),
+   None),
+  ('ageRange',
+   """
+   SELECT
+     CASE
+      WHEN YEARS_OLD[date_of_birth] < 0 THEN 'UNSET'
+      WHEN YEARS_OLD[date_of_birth] <= 17 THEN '0-17'
+      WHEN YEARS_OLD[date_of_birth] <= 25 THEN '18-25'
+      WHEN YEARS_OLD[date_of_birth] <= 35 THEN '26-35'
+      WHEN YEARS_OLD[date_of_birth] <= 45 THEN '36-45'
+      WHEN YEARS_OLD[date_of_birth] <= 55 THEN '46-55'
+      WHEN YEARS_OLD[date_of_birth] <= 65 THEN '56-65'
+      WHEN YEARS_OLD[date_of_birth] <= 75 THEN '66-75'
+      WHEN YEARS_OLD[date_of_birth] <= 85 THEN '76-85'
+      ELSE '86+'
+     END age_range,
+     SUM(1)
+   FROM participant_summary
+   GROUP BY 1;
+   """,
+   None),
+  ('physicalMeasurements',
+   """
+   SELECT physical_measurements_status, SUM(1)
+   FROM participant_summary
+   GROUP BY 1;
+   """,
+   lambda v: PhysicalMeasurementsStatus.lookup_by_number(v).name),
+  ('biospecimenSamples',
+   """
+   SELECT
+     CASE (biospecimen_status)
+      WHEN '{}' THEN 'UNSET'
+      WHEN '{}' THEN 'UNSET'
+      ELSE 'COLLECTED'
+     END, SUM(1)
+   FROM participant_summary
+   GROUP BY 1;
+   """.format(str(OrderStatus.UNSET.number), str(OrderStatus.CREATED.number)),
+   None),
+  _questionnaire_metric('questionnaireOnOverallHealth', 'questionnaire_on_overall_health'),
+  # Personal habits is a newer naming for lifestyle
+  _questionnaire_metric('questionnaireOnPersonalHabits', 'questionnaire_on_lifestyle'),
+  # Sociodemographics is a newer naming for 'the basics'
+  _questionnaire_metric('questionnaireOnSociodemographics', 'questionnaire_on_the_basics'),
+  _questionnaire_metric('questionnaireOnHealthcareAccess', 'questionnaire_on_healthcare_access'),
+  _questionnaire_metric('questionnaireOnMedicalHistory', 'questionnaire_on_medical_history'),
+  _questionnaire_metric('questionnaireOnMedications', 'questionnaire_on_medications'),
+  _questionnaire_metric('questionnaireOnFamilyHealth', 'questionnaire_on_family_health')
+]
+
+
+class PublicMetricsExport(object):
+  """Exports data from the database needed to generate public registration metrics."""
+
+  @staticmethod
+  def export():
+    # TODO(calbach): Write the output to a given target destination rather than
```

Comment:
It may make sense to have a method like this that generates the JSON we want to write out, and have something else that writes it to the database. (I'm not sure.)

Let's also talk about where these wind up.

Reply:
One reason to potentially gather all the metrics in a method like this and then write them in a second one: we can write out all the metrics in a single transaction, updating them in place, and have them be consistent, rather than having to worry about metric versioning like we do with the pipeline. (You could do that in the same transaction as the reads I guess, but it seems better to break them up.)